builtin/tfc/config_sourcer.go

```diff
@@ -0,0 +1,478 @@
+				},
+			}
+		}
+	}
+
+	baseURL := cs.config.BaseURL
+	if baseURL == "" {
+		baseURL = DefaultBaseURL
+	}
+
+	// If we have a last read value and its before our refresh period, we
+	// just returned cached values. So when the time is up, we just nuke
+	// the cache to force ourselves to recalculate.
+	if cs.lastRead.IsZero() || time.Now().Sub(cs.lastRead) > cs.refreshPeriod {
+		log.Trace("purging cached secrets")
+		cs.secretCache = nil
+		cs.lastRead = time.Now()
+	}
+
+	if cs.secretCache == nil {
+		cs.secretCache = map[string]*cachedSecret{}
+	}
+
+	// Go through each request and read it. The way this generally works:
+	// If the variable is not in our cache, we read it from TFC.
+	//
+	// If a config change happens, the ConfigSourcer contract states that
+	// Stop will be called. When Stop is called, we clear our full cache and
+	// stop all renewers.
+	//
+	// In most cases, this returns cached values due to the long
+	// default refresh period.
+	var results []*pb.ConfigSource_Value
+	for _, req := range reqs {
+		result := &pb.ConfigSource_Value{Name: req.Name}
+		results = append(results, result)
+
+		// Decode our configuration
+		var tfcReq reqConfig
+		if err := mapstructure.WeakDecode(req.Config, &tfcReq); err != nil {
+			result.Result = &pb.ConfigSource_Value_Error{
+				Error: status.New(codes.Aborted, err.Error()).Proto(),
+			}
+
+			continue
+		}
+
+		L := log.With("workspace", tfcReq.Workspace, "organization", tfcReq.Organization)
+
+		// We have to map the organization + workspace  to a workspace-id, so we do that first.
+		// the workspaceIds map is never cleared beacuse the configuration about which
+		// organization + workspace that is in use is static in the context of a config
+		// sourcer.
+		key := tfcReq.Organization + "/" + tfcReq.Workspace
+
+		id, ok := cs.workspaceIds[key]
+		if !ok {
+			L.Trace("querying TFC workspace id")
+
+			url := fmt.Sprintf("%s/api/v2/organizations/%s/workspaces/%s",
+				baseURL, tfcReq.Organization, tfcReq.Workspace)
+
+			req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+			if err != nil {
+				L.Error("error constructing request", "error", err)
+
+				result.Result = &pb.ConfigSource_Value_Error{
+					Error: status.New(codes.Aborted, err.Error()).Proto(),
+				}
+
+				continue
+			}
+
+			req.Header.Set("Authorization", "Bearer "+cs.config.Token)
+			req.Header.Set("Content-Type", "application/vnd.api+json")
+
+			resp, err := http.DefaultClient.Do(req)
+			if err != nil {
+				L.Error("error sending request for workspace info", "error", err)
+
+				result.Result = &pb.ConfigSource_Value_Error{
+					Error: status.New(codes.Aborted, err.Error()).Proto(),
+				}
+
+				continue
+			}
+
+			defer resp.Body.Close()
+
+			if resp.StatusCode != 200 {
+				L.Error("error reading workspace info", "status-code", resp.StatusCode)
+
+				result.Result = &pb.ConfigSource_Value_Error{
+					Error: status.New(codes.Aborted, "fucked").Proto(),
```

Comment:
We should change this to something else :sweat_smile: 
```suggestion
					Error: status.New(codes.Aborted, "failed to read workspace info").Proto(),
```